### PR TITLE
client-go: Make conditionalTransport a WrapperRoundTripper

### PR DESCRIPTION
**Pull request by apelisse**
**What this PR does / why we need it**:
Fixes broken GKE test: https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke/

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go [hunk 1]
```diff
@@ -306,3 +306,7 @@ func (t *conditionalTransport) RoundTrip(req *http.Request) (*http.Response, err
 
 	return res, nil
 }
+
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: type assertion wouldn't be a bad idea:
```
var _ = net.RoundTripperWrapper(&conditionalTransport{})
```

**Comment by liggitt** [comment]
all the client auth providers that implement WrapTransport should ensure they return a round tripper that implements this. a quick sweep shows azureRoundTripper, oidcAuthProvider, and tokenRoundTripper all have the same issue

**Comment by apelisse** [comment]
But it doesn't even fix the problem ... We just fail with another error.

**Comment by liggitt** [comment]
> But it doesn't even fix the problem ... It just fails with another error.

It's making progress, since we're actually getting an API response now:

> Error from server (Forbidden): pods "nginx" is forbidden: User "system:anonymous" cannot create pods/exec in the namespace "e2e-tests-kubectl-9gfmk".: "No policy matched.\nUnknown user \"system:anonymous\""

That error means the apiserver is not seeing authentication credentials on the request. Looking at the upgrade proxy path in kubectl, I'm not actually seeing where the upgrade transport gets to handle the request to do things like add bearer tokens

**Comment by liggitt** [comment]
the issue is with https://github.com/kubernetes/kubernetes/blob/4bfe9b1a568be137f72cf5dfc8bc2c1f13874548/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L269-L282 that does not actually use the transport to handle the request, only to dial. That means that TLS-authentication methods like client certs work, and header authentication methods like bearer tokens do not.

**Comment by apelisse** [comment]
cc @liggitt

**Comment by liggitt** [comment]
> all the client auth providers that implement WrapTransport should ensure they return a round tripper that implements this. a quick sweep shows azureRoundTripper, oidcAuthProvider, and tokenRoundTripper all have the same issue

you mind updating this to fix those as well, with type assertions?

**Comment by apelisse** [comment]
Yeah, I'll send an update very soon

**Comment by apelisse** [comment]
PTAL @lig

**Comment by liggitt** [comment]
/lgtm
thanks

**Pull request merged**